presto-hive/src/main/java/com/facebook/presto/hive/metastore/RecordingHiveMetastore.java

```diff
@@ -0,0 +1,582 @@
+        }
+
+        return CacheBuilder.<K, V>newBuilder()
+                .expireAfterWrite(hiveClientConfig.getRecordingDuration().toMillis(), MILLISECONDS)
+                .build();
+    }
+
+    @Managed
+    public void writeRecording()
+            throws IOException
+    {
+        if (replay) {
+            throw new PrestoException(CONFIGURATION_INVALID, "Cannot write recording in replay mode");
+        }
+
+        Recording recording = new Recording(
+                allDatabases,
+                toPairsList(databaseCache),
+                toPairsList(tableCache),
+                toPairsList(tableStatisticsCache),
+                toPairsList(partitionStatisticsCache),
+                toPairsList(allTablesCache),
+                toPairsList(allViewsCache),
+                toPairsList(partitionCache),
+                toPairsList(partitionNamesCache),
+                toPairsList(partitionNamesByPartsCache),
+                toPairsList(partitionsByNamesCache),
+                toPairsList(rolesCache),
+                toPairsList(databasePrivilegesCache),
+                toPairsList(tablePrivilegesCache));
+        new ObjectMapperProvider().get()
+                .writerWithDefaultPrettyPrinter()
+                .writeValue(new File(recordingPath), recording);
+    }
+
+    private static <K, V> Map<K, V> toMap(List<Pair<K, V>> pairs)
+    {
+        return pairs.stream()
+                .collect(ImmutableMap.toImmutableMap(Pair::getKey, Pair::getValue));
+    }
+
+    private static <K, V> List<Pair<K, V>> toPairsList(Cache<K, V> cache)
+    {
+        return cache.asMap().entrySet().stream()
+                .map(entry -> new Pair<>(entry.getKey(), entry.getValue()))
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public Optional<Database> getDatabase(String databaseName)
+    {
+        return loadValue(databaseCache, databaseName, () -> delegate.getDatabase(databaseName));
+    }
+
+    @Override
+    public List<String> getAllDatabases()
+    {
+        if (replay) {
+            return allDatabases.orElseGet(delegate::getAllDatabases);
+        }
+
+        List<String> result = delegate.getAllDatabases();
+        allDatabases = Optional.of(result);
+        return result;
+    }
+
+    @Override
+    public Optional<Table> getTable(String databaseName, String tableName)
+    {
+        return loadValue(tableCache, table(databaseName, tableName), () -> delegate.getTable(databaseName, tableName));
+    }
+
+    @Override
+    public Set<ColumnStatisticType> getSupportedColumnStatistics(Type type)
+    {
+        return delegate.getSupportedColumnStatistics(type);
+    }
+
+    @Override
+    public PartitionStatistics getTableStatistics(String databaseName, String tableName)
+    {
+        return loadValue(
+                tableStatisticsCache,
+                table(databaseName, tableName),
+                () -> delegate.getTableStatistics(databaseName, tableName));
```

Comment:
It would be slightly less err prone if the lambda accepted the cache key, the one that you construct in the preceding line

```suggestion
                key -> delegate.getTableStatistics(key.getDatabaseName(), key.getTableName()));
```

Reply:
I think that's unneeded complexity

Reply:
I think some more complexity is warranted, if this makes code harder to break/to make mistake. 

up to you